r/Dockerfile

```diff
@@ -73,14 +64,16 @@ ENV RENV_LIB=/src/lib
 RUN echo ".libPaths(c('$RENV_LIB', .libPaths()))" >> $(R RHOME)/etc/Rprofile.site
 
 # install runtime system deps
+# python3-dev prevents 'reticulate can only bind to copies of Python built with --enable-shared'
 # cleanup setup files
-RUN echo "python3-pip" >> sysdeps_run.txt && \
+RUN echo "python3-pip python3-venv python3-dev" >> sysdeps_run.txt && \
```

Comment:
why not declare these in the `sysdeps_build_debian.txt` instead of the dynamically generated `sysdeps_run.txt`?

Reply:
`python3-pip` is actually already installed

https://github.com/hms-dbmi-cellenics/worker/blob/f07d5a1bcdebeb70c860beef324c1a11952ffcb8/r/setup/sysdeps_build_debian.txt#L4C1-L5C1

Reply:
`sysdeps_build_debian.txt` are dependencies for building the R-packages. Once they have been built, the run-time system dependencies for R packages are determined  by `get_sysdeps_run.R`. There are however some additional dependencies that are necessary for non R-package things:

- python3-pip for the development build (not needed for prod but installed here for code simplicity)
- python3-venv to setup the venv for `umap-learn`
- python3-dev to have a copy of python that `reticulate` can use

Reply:
I tried it in the pipeline PR and it does not work, for some reason I don't understand.

Reply:
Not tested but builds for me: https://github.com/hms-dbmi-cellenics/pipeline/pull/340

Reply:
Thanks, I'll test the pipeline PR.

Regarding build time dependencies, are they removed downstream, after building the R packages?That relates to my initial question, why do we need python3-pip installed twice?

Reply:
Yes the `common` build doesn't start from the end of the `builder`, but rather from a fresh `rocker/r-ver`. It transfers over already built packages etc (`COPY --from=builder /src /src`) and then installs only the system dependencies that are needed to **run** packages (less hefty that packages needed to **build** all the R/python packages).